File: tree/easy/first_bad_version_278.go
package easy

/**
 * Forward declaration of isBadVersion API.
 * @param   version   your guess about first bad version
 * @return 	 	      true if current version is bad
 *			          false if current version is good
 * func isBadVersion(version int) bool;
 */

func isBadVersion(version int) bool {
	return version >= 1
}

func firstBadVersion(n int) int {
	left, right := 0, n
	for left <= right {
		middle := (left + right) / 2
		if !isBadVersion(middle-1) && isBadVersion(middle) {
			return middle
		}
		if isBadVersion(middle) {
			right = middle - 1
		} else {
			left = middle + 1
		}
	}
	return -1
}

File: tree/easy/java_solutions/FirstBadVersion_278.java
package tree.easy.java_solutions;

public class FirstBadVersion_278 {

    private record Solution(int[] versions) {

        public int firstBadVersion(int n) {
            int left = 0, right = n;
            while (left <= right) {
                int middle = left + (right - left) / 2;
                if (!isBadVersion(middle - 1) && isBadVersion(middle)) {
                    return middle;
                }
                if (isBadVersion(middle)) {
                    right = middle - 1;
                } else {
                    left = middle + 1;
                }
            }
            return -1;
        }

        private boolean isBadVersion(int idx) {
            return versions[idx] == -1;
        }
    }
}
